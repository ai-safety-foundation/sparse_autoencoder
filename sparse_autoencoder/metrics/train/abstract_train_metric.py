"""Abstract train metric."""
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any

from sparse_autoencoder.tensor_types import (
    InputOutputActivationBatch,
    LearnedActivationBatch,
)


@dataclass
class TrainMetricData:
    """Train metric data."""

    input_activations: InputOutputActivationBatch

    learned_activations: LearnedActivationBatch

    decoded_activations: InputOutputActivationBatch


class AbstractTrainMetric(ABC):
    """Abstract train metric."""

    @abstractmethod
    def calculate(self, data: TrainMetricData) -> dict[str, Any]:
        """Calculate any metrics.

        Args:
            data: Train metric data.

        Returns:
            Dictionary of metrics.
        """
