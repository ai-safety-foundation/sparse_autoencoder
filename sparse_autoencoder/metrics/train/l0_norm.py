"""L0 norm sparsity metric."""
from typing import Any

from jaxtyping import Float, Int64
from pydantic import PositiveInt, validate_call
import torch
from torch import Tensor
from torchmetrics import Metric

from sparse_autoencoder.tensor_types import Axis
from sparse_autoencoder.utils.tensor_shape import shape_with_optional_dimensions


class L0NormMetric(Metric):
    """Learned activations L0 norm metric.

    The L0 norm is the number of non-zero elements in a learned activation vector, averaged over the
    number of activation vectors.

    Examples:
        >>> metric = L0NormMetric()
        >>> learned_activations = torch.tensor([
        ...     [1., 0., 1.], # Batch 1 (single component): learned features (2 active neurons)
        ...     [0., 1., 0.]  # Batch 2 (single component): learned features (1 active neuron)
        ... ])
        >>> metric.forward(learned_activations)
        tensor(1.5000)

        With 2 components, the metric will return the average number of active (non-zero)
        neurons as a 1d tensor.

        >>> metric = L0NormMetric(num_components=2)
        >>> learned_activations = torch.tensor([
        ...     [ # Batch 1
        ...         [1., 0., 1.], # Component 1: learned features (2 active neurons)
        ...         [1., 0., 1.]  # Component 2: learned features (2 active neurons)
        ...     ],
        ...     [ # Batch 2
        ...         [0., 1., 0.], # Component 1: learned features (1 active neuron)
        ...         [1., 0., 1.]  # Component 2: learned features (2 active neurons)
        ...     ]
        ... ])
        >>> metric.forward(learned_activations)
        tensor([1.5000, 2.0000])
    """

    # Torchmetrics settings
    is_differentiable: bool | None = False
    full_state_update: bool | None = False
    plot_lower_bound: float | None = 0.0

    # State
    active_neurons_count: Float[Tensor, Axis.COMPONENT_OPTIONAL]
    num_activation_vectors: Int64[Tensor, Axis.SINGLE_ITEM]

    @validate_call
    def __init__(self, num_components: PositiveInt | None = None) -> None:
        """Initialize the metric."""
        super().__init__()

        self.add_state(
            "active_neurons_count",
            default=torch.zeros(shape_with_optional_dimensions(num_components), dtype=torch.float),
            dist_reduce_fx="sum",  # Float is needed for dist reduce to work
        )

        self.add_state(
            "num_activation_vectors",
            default=torch.tensor(0, dtype=torch.int64),
            dist_reduce_fx="sum",
        )

    def update(
        self,
        learned_activations: Float[
            Tensor, Axis.names(Axis.PROCESS_BATCH, Axis.COMPONENT_OPTIONAL, Axis.LEARNT_FEATURE)
        ],
        **kwargs: Any,  # type: ignore # noqa: ARG002, ANN401 (allows combining with other metrics)
    ) -> None:
        """Update the metric state.

        Args:
            learned_activations: The learned activations.
            **kwargs: Ignored keyword arguments (to allow use with other metrics in a collection).
        """
        self.num_activation_vectors += learned_activations.shape[0]

        self.active_neurons_count += torch.count_nonzero(learned_activations, dim=-1).sum(
            dim=0, dtype=torch.int64
        )

    def compute(
        self,
    ) -> Float[Tensor, Axis.names(Axis.COMPONENT_OPTIONAL)]:
        """Compute the metric.

        Note that torchmetrics converts shape `[0]` tensors into scalars (shape `0`).
        """
        return self.active_neurons_count / self.num_activation_vectors
