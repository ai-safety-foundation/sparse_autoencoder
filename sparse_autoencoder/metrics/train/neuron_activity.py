"""Neuron activity metric."""
from typing import Annotated, Any

from jaxtyping import Bool, Float, Int64
from pydantic import Field, NonNegativeFloat, PositiveInt, validate_call
import torch
from torch import Tensor
from torchmetrics import Metric

from sparse_autoencoder.tensor_types import Axis
from sparse_autoencoder.utils.tensor_shape import shape_with_optional_dimensions


class NeuronActivityMetric(Metric):
    """Neuron activity metric.

    Example:
        With a single component and a horizon of 2 activations, the metric will return nothing
        after the first activation is added and then computed, and then return the number of dead
        neurons after the second activation is added (with update). The breakdown by component isn't
        shown here as there is just one component.

        >>> metric = NeuronActivityMetric(num_learned_features=3)
        >>> learned_activations = torch.tensor([
        ...     [1., 0., 1.], # Batch 1 (single component): learned features (2 active neurons)
        ...     [0., 0., 0.]  # Batch 2 (single component): learned features (0 active neuron)
        ... ])
        >>> metric.forward(learned_activations)
        tensor(1)
    """

    # Torchmetrics settings
    is_differentiable: bool | None = False
    full_state_update: bool | None = True
    plot_lower_bound: float | None = 0.0

    # Metric settings
    _threshold_is_dead_portion_fires: NonNegativeFloat

    # State
    neuron_fired_count: Float[Tensor, Axis.names(Axis.COMPONENT_OPTIONAL, Axis.LEARNT_FEATURE)]
    num_activation_vectors: Int64[Tensor, Axis.SINGLE_ITEM]

    @validate_call
    def __init__(
        self,
        num_learned_features: PositiveInt,
        num_components: PositiveInt | None = None,
        threshold_is_dead_portion_fires: Annotated[float, Field(strict=True, ge=0, le=1)] = 0.0,
    ) -> None:
        """Initialise the metric.

        Args:
            num_learned_features: Number of learned features.
            num_components: Number of components.
            threshold_is_dead_portion_fires: Thresholds for counting a neuron as dead (portion of
                activation vectors that it fires for must be less than or equal to this number).
                Commonly used values are 0.0, 1e-5 and 1e-6.
        """
        super().__init__()
        self._threshold_is_dead_portion_fires = threshold_is_dead_portion_fires

        self.add_state(
            "neuron_fired_count",
            default=torch.zeros(
                shape_with_optional_dimensions(num_components, num_learned_features),
                dtype=torch.float,  # Float is needed for dist reduce to work
            ),
            dist_reduce_fx="sum",
        )

        self.add_state(
            "num_activation_vectors",
            default=torch.tensor(0, dtype=torch.int64),
            dist_reduce_fx="sum",
        )

    def update(
        self,
        learned_activations: Float[
            Tensor, Axis.names(Axis.PROCESS_BATCH, Axis.COMPONENT_OPTIONAL, Axis.LEARNT_FEATURE)
        ],
        **kwargs: Any,  # type: ignore # noqa: ARG002, ANN401 (allows combining with other metrics)
    ) -> None:
        """Update the metric state.

        Args:
            learned_activations: The learned activations.
            **kwargs: Ignored keyword arguments (to allow use with other metrics in a collection).
        """
        # Increment the counter of activations seen since the last compute step
        self.num_activation_vectors += learned_activations.shape[0]

        # Count the number of active neurons in the batch
        neuron_has_fired: Bool[
            Tensor, Axis.names(Axis.PROCESS_BATCH, Axis.COMPONENT_OPTIONAL, Axis.LEARNT_FEATURE)
        ] = torch.gt(learned_activations, 0)

        self.neuron_fired_count += neuron_has_fired.sum(dim=0, dtype=torch.float)

    def compute(self) -> Int64[Tensor, Axis.COMPONENT_OPTIONAL]:
        """Compute the metric.

        Note that torchmetrics converts shape `[0]` tensors into scalars (shape `0`).
        """
        threshold_activations: Float[Tensor, Axis.SINGLE_ITEM] = (
            self._threshold_is_dead_portion_fires * self.num_activation_vectors
        )

        return torch.sum(
            self.neuron_fired_count <= threshold_activations, dim=-1, dtype=torch.int64
        )
