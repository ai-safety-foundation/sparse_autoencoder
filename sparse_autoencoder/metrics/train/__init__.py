"""Train step metrics."""
from sparse_autoencoder.metrics.train.abstract_train_metric import AbstractTrainMetric


__all__ = ["AbstractTrainMetric"]
