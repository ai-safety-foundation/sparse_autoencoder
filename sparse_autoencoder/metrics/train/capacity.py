"""Capacity Metrics."""
from typing import Any

import einops
import numpy as np
from numpy import histogram
from numpy.typing import NDArray
import torch
import wandb

from sparse_autoencoder.metrics.train.abstract_train_metric import (
    AbstractTrainMetric,
    TrainMetricData,
)
from sparse_autoencoder.tensor_types import LearnedActivationBatch, TrainBatchStatistic


class CapacityMetric(AbstractTrainMetric):
    """Capacities Metrics for Learned Features.

    Measure the capacity of a set of features as defined in [Polysemanticity and Capacity in Neural Networks](https://arxiv.org/pdf/2210.01892.pdf).

    Capacity is intuitively measuring the 'proportion of a dimension' assigned to a feature.
    Formally it's the ratio of the squared dot product of a feature with itself to the sum of its
    squared dot products of all features.

    If the features are orthogonal, the capacity is 1. If they are all the same, the capacity is
    1/n.
    """

    @staticmethod
    def capacities(features: LearnedActivationBatch) -> TrainBatchStatistic:
        """Calculate capacities.

        Example:
            >>> import torch
            >>> orthogonal_features = torch.tensor([[1., 0., 0.], [0., 1., 0.], [0., 0., 1.]])
            >>> orthogonal_caps = CapacityMetric.capacities(orthogonal_features)
            >>> orthogonal_caps
            tensor([1., 1., 1.])

        Args:
            features: A collection of features.

        Returns:
            A 1D tensor of capacities, where each element is the capacity of the corresponding
            feature.
        """
        squared_dot_products = (
            einops.einsum(
                features, features, "n_feats1 feat_dim, n_feats2 feat_dim -> n_feats1 n_feats2"
            )
            ** 2
        )
        sum_of_sq_dot = squared_dot_products.sum(dim=-1)
        return torch.diag(squared_dot_products) / sum_of_sq_dot

    @staticmethod
    def wandb_capacities_histogram(
        capacities: TrainBatchStatistic,
    ) -> wandb.Histogram:
        """Create a W&B histogram of the capacities.

        This can be logged with Weights & Biases using e.g. `wandb.log({"capacities_histogram":
        wandb_capacities_histogram(capacities)})`.

        Args:
            capacities: Capacity of each feature. Can be calculated using :func:`calc_capacities`.

        Returns:
            Weights & Biases histogram for logging with `wandb.log`.
        """
        numpy_capacities: NDArray[np.float_] = capacities.detach().cpu().numpy()

        bins, values = histogram(numpy_capacities, bins=20, range=(0, 1))
        return wandb.Histogram(np_histogram=(bins, values))

    def calculate(self, data: TrainMetricData) -> dict[str, Any]:
        """Calculate the capacities for a training batch."""
        train_batch_capacities = self.capacities(data.learned_activations)
        train_batch_capacities_histogram = self.wandb_capacities_histogram(train_batch_capacities)

        return {
            "train_batch_capacities_histogram": train_batch_capacities_histogram,
        }
