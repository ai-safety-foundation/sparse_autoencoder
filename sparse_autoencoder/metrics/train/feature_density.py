"""Train batch feature density."""
from typing import Any

import einops
import numpy as np
from numpy import histogram
from numpy.typing import NDArray
import torch
import wandb

from sparse_autoencoder.metrics.train.abstract_train_metric import (
    AbstractTrainMetric,
    TrainMetricData,
)
from sparse_autoencoder.tensor_types import LearnedActivationBatch, LearntActivationVector


class TrainBatchFeatureDensityMetric(AbstractTrainMetric):
    """Train batch feature density.

    Percentage of samples in which each feature was active (i.e. the neuron has "fired"), in a
    training batch.

    Generally we want a small number of features to be active in each batch, so average feature
    density should be low. By contrast if the average feature density is high, it means that the
    features are not sparse enough.

    Warning:
        This is not the same as the feature density of the entire training set. It's main use is
        tracking the progress of training.
    """

    threshold: float

    def __init__(self, threshold: float = 0.0) -> None:
        """Initialise the train batch feature density metric.

        Args:
            threshold: Threshold for considering a feature active (i.e. the neuron has "fired").
                This should be close to zero.
        """
        super().__init__()
        self.threshold = threshold

    def feature_density(self, activations: LearnedActivationBatch) -> LearntActivationVector:
        """Count how many times each feature was active.

        Percentage of samples in which each feature was active (i.e. the neuron has "fired").

        Example:
            >>> import torch
            >>> activations = torch.tensor([[0.5, 0.5, 0.0], [0.5, 0.0, 0.0001]])
            >>> TrainBatchFeatureDensityMetric(0.001).feature_density(activations).tolist()
            [1.0, 0.5, 0.0]

        Args:
            activations: Sample of cached activations (the Autoencoder's learned features).

        Returns:
            Number of times each feature was active in a sample.
        """
        has_fired: LearnedActivationBatch = torch.gt(activations, self.threshold).to(
            dtype=torch.float  # Move to float so it can be averaged
        )

        return einops.reduce(has_fired, "sample activation -> activation", "mean")

    @staticmethod
    def wandb_feature_density_histogram(
        feature_density: LearntActivationVector,
    ) -> wandb.Histogram:
        """Create a W&B histogram of the feature density.

        This can be logged with Weights & Biases using e.g. `wandb.log({"feature_density_histogram":
        wandb_feature_density_histogram(feature_density)})`.

        Args:
            feature_density: Number of times each feature was active in a sample. Can be calculated
                using :func:`feature_activity_count`.

        Returns:
            Weights & Biases histogram for logging with `wandb.log`.
        """
        numpy_feature_density: NDArray[np.float_] = feature_density.detach().cpu().numpy()

        bins, values = histogram(numpy_feature_density, bins=50)
        return wandb.Histogram(np_histogram=(bins, values))

    def calculate(self, data: TrainMetricData) -> dict[str, Any]:
        """Calculate the train batch feature density metrics.

        Args:
            data: Train metric data.

        Returns:
            Dictionary with the train batch feature density metric, and a histogram of the feature
            density.
        """
        train_batch_feature_density: LearntActivationVector = self.feature_density(
            data.learned_activations
        )

        train_batch_feature_density_histogram: wandb.Histogram = (
            self.wandb_feature_density_histogram(train_batch_feature_density)
        )

        return {
            "train_batch_feature_density_histogram": train_batch_feature_density_histogram,
        }
