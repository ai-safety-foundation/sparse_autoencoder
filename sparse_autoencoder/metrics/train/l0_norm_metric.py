"""L0 norm sparsity metric."""
from typing import final

import torch

from sparse_autoencoder.metrics.train.abstract_train_metric import (
    AbstractTrainMetric,
    TrainMetricData,
)


@final
class TrainBatchLearnedActivationsL0(AbstractTrainMetric):
    """Learned activations L0 norm sparsity metric.

    The L0 norm is the number of non-zero elements in a learned activation vector. We then average
    this over the batch.
    """

    def calculate(self, data: TrainMetricData) -> dict[str, float]:
        """Create a log item for Weights and Biases."""
        batch_size = data.learned_activations.size(0)
        n_non_zero_activations = torch.count_nonzero(data.learned_activations)
        batch_average = n_non_zero_activations / batch_size
        return {"learned_activations_l0_norm": batch_average.item()}
