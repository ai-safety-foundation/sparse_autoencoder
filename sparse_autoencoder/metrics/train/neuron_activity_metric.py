"""Neuron activity metric."""
from typing import TYPE_CHECKING, Any

import numpy as np
from numpy.typing import NDArray
import wandb

from sparse_autoencoder.metrics.train.abstract_train_metric import (
    AbstractTrainMetric,
    TrainMetricData,
)


if TYPE_CHECKING:
    from sparse_autoencoder.tensor_types import Axis
from jaxtyping import Int
from torch import Tensor


class NeuronActivityMetric(AbstractTrainMetric):
    """Neuron activity metric."""

    def __init__(self, horizon: int = 100) -> None:
        """Initialise the neuron activity metric.

        Args:
            horizon: Number of batches to average over.
        """
        super().__init__()
        self.horizon = horizon
        self.neuron_activity: None | Int[Tensor, Axis.LEARNT_FEATURE] = None
        self.batch_count = 0
        self.activation_eps = 0.1

    def calculate(self, data: TrainMetricData) -> dict[str, Any]:
        """Calculate the neuron activity metrics.

        Args:
            data: Resample metric data.

        Returns:
            Dictionary of metrics.
        """
        fired = data.learned_activations > 0
        if self.batch_count == 0:
            self.neuron_activity = fired.sum(dim=0)
        else:
            self.neuron_activity += fired.sum(dim=0)

        self.batch_count += 1

        if self.batch_count >= self.horizon:
            # Histogram of neuron activity
            numpy_neuron_activity: NDArray[np.float_] = self.neuron_activity.detach().cpu().numpy()
            log_neuron_activity = np.log(numpy_neuron_activity + self.activation_eps)
            bins, values = np.histogram(numpy_neuron_activity, bins=50)
            histogram = wandb.Histogram(np_histogram=(bins, values))
            log_bins, log_values = np.histogram(log_neuron_activity, bins=50)
            log_histogram = wandb.Histogram(np_histogram=(log_bins, log_values))

            self.batch_count = 0
            return {
                "alive_neuron_count": (self.neuron_activity > 0).sum().item(),
                "dead_neuron_count": (self.neuron_activity == 0).sum().item(),
                "neuron_activity_histogram": histogram,
                "log_neuron_activity_histogram": log_histogram,
            }
        return {}
