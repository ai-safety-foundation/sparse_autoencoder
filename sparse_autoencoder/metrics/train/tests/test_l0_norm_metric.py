"""Tests for the L0NormMetric class."""
import torch

from sparse_autoencoder.metrics.train.abstract_train_metric import TrainMetricData
from sparse_autoencoder.metrics.train.l0_norm_metric import TrainBatchLearnedActivationsL0


def test_l0_norm_metric() -> None:
    """Test the L0NormMetric."""
    learned_activations = torch.tensor([[1.0, 0.0, 0.0], [0.0, 0.01, 2.0]])
    l0_norm_metric = TrainBatchLearnedActivationsL0()
    data = TrainMetricData(
        input_activations=torch.zeros_like(learned_activations),
        learned_activations=learned_activations,
        decoded_activations=torch.zeros_like(learned_activations),
    )
    log = l0_norm_metric.calculate(data)
    expected = 3 / 2
    assert log["learned_activations_l0_norm"] == expected
