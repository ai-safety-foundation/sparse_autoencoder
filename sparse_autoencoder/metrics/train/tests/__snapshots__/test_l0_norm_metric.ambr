# serializer version: 1
# name: test_weights_biases_log_matches_snapshot
  list([
    dict({
      'component_0/train/learned_activations_l0_norm': tensor(8.),
      'component_1/train/learned_activations_l0_norm': tensor(8.),
      'component_2/train/learned_activations_l0_norm': tensor(8.),
      'component_3/train/learned_activations_l0_norm': tensor(8.),
      'component_4/train/learned_activations_l0_norm': tensor(8.),
      'component_5/train/learned_activations_l0_norm': tensor(8.),
      'train/learned_activations_l0_norm/component_mean': tensor(8.),
    }),
  ])
# ---
